addons/website_slides/models/slide_slide.py

```diff
@@ -854,13 +883,18 @@ def _action_set_viewed(self, target_partner, quiz_attempts_inc=False):
             'quiz_attempts_count': 1 if quiz_attempts_inc else 0,
             'vote': 0} for new_slide in new_slides])
 
-    def action_set_completed(self):
-        if any(not slide.channel_id.is_member for slide in self):
+    def action_mark_completed(self):
+        if any(not slide.can_self_mark_completed for slide in self):
             raise UserError(_('You cannot mark a slide as completed if you are not among its members.'))
 
-        return self._action_set_completed(self.env.user.partner_id)
+        return self._action_mark_completed()
 
-    def _action_set_completed(self, target_partner):
+    def _action_mark_completed(self):
+        if any(slide.user_has_completed for slide in self):
+            raise UserError(_('This slide is already completed.'))
```

Comment:
Should we simply filter-out slides that are already completed?

Raising an error seems harsh.